File: server/internal/queue/queue.go
// Package queue runs build jobs on Asynq and keeps the ticket numbers,
// queue positions and wait estimates the front end shows (plan.md 1.8).
package queue

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strconv"
	"time"

	"github.com/hibiken/asynq"
	"github.com/redis/go-redis/v9"
)

// Classes, by what was asked for; Asynq serves them with these priorities.
var Classes = map[string]int{"record": 6, "build": 3, "pack": 1}

const TaskType = "ib:job"

// Job is what's stored per job in Redis (a JSON value).
type Job struct {
	ID       string          `json:"id"`
	Ticket   int64           `json:"ticket"`
	Class    string          `json:"class"`
	Status   string          `json:"status"` // queued, running, done, failed
	Progress string          `json:"progress"`
	Error    *string         `json:"error"`
	Request  json.RawMessage `json:"request"`
	Result   json.RawMessage `json:"result,omitempty"`
	Created  time.Time       `json:"created"`
	Started  *time.Time      `json:"started,omitempty"`
	Finished *time.Time      `json:"finished,omitempty"`
}

type Queue struct {
	rdb    *redis.Client
	client *asynq.Client
	opt    asynq.RedisClientOpt
	// Workers per class, for the ETA.
	Workers int
}

func New(addr string, db, workers int) *Queue {
	opt := asynq.RedisClientOpt{Addr: addr, DB: db}
	return &Queue{rdb: redis.NewClient(&redis.Options{Addr: addr, DB: db}), client: asynq.NewClient(opt), opt: opt, Workers: workers}
}

func (q *Queue) Ping(ctx context.Context) error { return q.rdb.Ping(ctx).Err() }

func key(parts ...string) string {
	k := "ib"
	for _, p := range parts {
		k += ":" + p
	}
	return k
}

func newID() string {
	b := make([]byte, 9)
	rand.Read(b)
	return "j_" + hex.EncodeToString(b)
}

// Submit queues a job and returns it with its ticket number.
func (q *Queue) Submit(ctx context.Context, class string, req json.RawMessage) (*Job, error) {
	if _, ok := Classes[class]; !ok {
		return nil, fmt.Errorf("unknown class %q", class)
	}
	t, err := q.rdb.Incr(ctx, key("ticket")).Result()
	if err != nil {
		return nil, err
	}
	j := &Job{ID: newID(), Ticket: t, Class: class, Status: "queued", Progress: "Waiting in the queue", Request: req, Created: time.Now().UTC()}
	if err := q.Save(ctx, j); err != nil {
		return nil, err
	}
	if err := q.rdb.ZAdd(ctx, key("queued", class), redis.Z{Score: float64(t), Member: j.ID}).Err(); err != nil {
		return nil, err
	}
	payload, _ := json.Marshal(map[string]string{"id": j.ID})
	_, err = q.client.EnqueueContext(ctx, asynq.NewTask(TaskType, payload), asynq.Queue(class),
		asynq.MaxRetry(0), asynq.Timeout(2*time.Hour), asynq.Retention(24*time.Hour))
	if err != nil {
		q.rdb.ZRem(ctx, key("queued", class), j.ID)
		return nil, err
	}
	return j, nil
}

func (q *Queue) Save(ctx context.Context, j *Job) error {
	b, _ := json.Marshal(j)
	return q.rdb.Set(ctx, key("job", j.ID), b, 7*24*time.Hour).Err()
}

func (q *Queue) Get(ctx context.Context, id string) (*Job, error) {
	b, err := q.rdb.Get(ctx, key("job", id)).Bytes()
	if err != nil {
		return nil, err
	}
	j := &Job{}
	return j, json.Unmarshal(b, j)
}

// Position is how many tickets of the same class are queued ahead.
func (q *Queue) Position(ctx context.Context, j *Job) int64 {
	if j.Status != "queued" {
		return 0
	}
	r, err := q.rdb.ZRank(ctx, key("queued", j.Class), j.ID).Result()
	if err != nil {
		return 0
	}
	return r
}

// ETA estimates seconds until the job finishes: the jobs ahead plus this
// one, at the recent average duration, spread over the workers. nil when
// there's no history yet.
func (q *Queue) ETA(ctx context.Context, j *Job, ahead int64) *int64 {
	durs, err := q.rdb.LRange(ctx, key("dur", j.Class), 0, 49).Result()
	if err != nil || len(durs) == 0 {
		return nil
	}
	var sum float64
	for _, d := range durs {
		f, _ := strconv.ParseFloat(d, 64)
		sum += f
	}
	avg := sum / float64(len(durs))
	w := q.Workers
	if w < 1 {
		w = 1
	}
	var secs float64
	switch j.Status {
	case "queued":
		running, _ := q.rdb.SCard(ctx, key("running", j.Class)).Result()
		secs = (float64(ahead+running)/float64(w) + 1) * avg
	case "running":
		if j.Started != nil {
			secs = avg - time.Since(*j.Started).Seconds()
		}
		if secs < 1 {
			secs = 1
		}
	default:
		return nil
	}
	n := int64(secs + 0.5)
	return &n
}

// Depths returns queued jobs per class.
func (q *Queue) Depths(ctx context.Context) map[string]int64 {
	out := map[string]int64{}
	for c := range Classes {
		out[c], _ = q.rdb.ZCard(ctx, key("queued", c)).Result()
	}
	return out
}

// Handler does the work for one job; it reports progress through set.
type Handler func(ctx context.Context, j *Job, progress func(string)) (json.RawMessage, error)

// Serve runs the workers until ctx ends.
func (q *Queue) Serve(ctx context.Context, h Handler) error {
	srv := asynq.NewServer(q.opt, asynq.Config{Concurrency: q.Workers, Queues: Classes,
		Logger: nil, LogLevel: asynq.WarnLevel})
	mux := asynq.NewServeMux()
	mux.HandleFunc(TaskType, func(tctx context.Context, t *asynq.Task) error {
		var p struct{ ID string }
		if err := json.Unmarshal(t.Payload(), &p); err != nil {
			return err
		}
		j, err := q.Get(tctx, p.ID)
		if err != nil {
			return err
		}
		now := time.Now().UTC()
		j.Status, j.Started, j.Progress = "running", &now, "Starting"
		q.rdb.ZRem(tctx, key("queued", j.Class), j.ID)
		q.rdb.SAdd(tctx, key("running", j.Class), j.ID)
		defer q.rdb.SRem(context.Background(), key("running", j.Class), j.ID)
		q.Save(tctx, j)
		res, herr := h(tctx, j, func(s string) {
			j.Progress = s
			q.Save(context.Background(), j)
		})
		end := time.Now().UTC()
		j.Finished = &end
		if herr != nil {
			msg := herr.Error()
			j.Status, j.Error, j.Progress = "failed", &msg, "Failed"
		} else {
			j.Status, j.Result, j.Progress = "done", res, "Done"
			d := end.Sub(*j.Started).Seconds()
			q.rdb.LPush(context.Background(), key("dur", j.Class), strconv.FormatFloat(d, 'f', 2, 64))
			q.rdb.LTrim(context.Background(), key("dur", j.Class), 0, 49)
		}
		q.Save(context.Background(), j)
		return nil // failures are reported on the job, not retried
	})
	go func() {
		<-ctx.Done()
		srv.Shutdown()
	}()
	return srv.Run(mux)
}

File: server/lib/queue.js
// The job queue (plan.md 1.8): BullMQ on Redis runs the work; the ticket
// numbers, queue positions and wait estimates the front end shows are kept
// in Redis under the Go server's keys (ib:ticket, ib:job:<id>,
// ib:queued:<class>, ib:running:<class>, ib:dur:<class>), so the job JSON
// and the numbers mean the same.
//
// Classes, by what was asked for: `record` (mode A) before `build` (modes B
// and C) on the shared workers (BullMQ priorities, strict where Asynq's
// were weighted), and `pack` (offline installers: downloads) on a worker of
// its own, one at a time, so packs can't hold up records and builds.
import crypto from 'node:crypto';
import { Queue, Worker } from 'bullmq';
import IORedis from 'ioredis';

export const CLASSES = { record: 1, build: 2, pack: 3 };   // BullMQ priority: 1 runs first
const PREFIX = 'ib-bull';
const JOBS = 'ib-jobs', PACKS = 'ib-pack';
const WEEK = 7 * 24 * 3600;
const TIMEOUT = 2 * 3600 * 1000;   // Asynq's task timeout on the Go server

const key = (...parts) => ['ib', ...parts].join(':');

function newID() {
  return 'j_' + crypto.randomBytes(9).toString('hex');
}

const now = () => new Date().toISOString();

export class JobQueue {
  // addr "host:port", db, workers
  constructor(addr, db, workers) {
    const i = addr.lastIndexOf(':');
    this.conn = { host: addr.slice(0, i) || '127.0.0.1', port: Number(addr.slice(i + 1)), db };
    this.workers = workers;
    this.rdb = new IORedis({ ...this.conn, maxRetriesPerRequest: 1, lazyConnect: false });
    this.rdb.on('error', () => { /* reported by ping and submit */ });
    const connection = { ...this.conn, maxRetriesPerRequest: null };
    this.queues = { [JOBS]: new Queue(JOBS, { connection, prefix: PREFIX }), [PACKS]: new Queue(PACKS, { connection, prefix: PREFIX }) };
    for (const q of Object.values(this.queues)) q.on('error', () => {});
    this.running = [];
  }

  async ping() {
    let t;
    try {
      await Promise.race([this.rdb.ping(), new Promise((_, j) => { t = setTimeout(() => j(new Error('timeout')), 2000); })]);
    } finally {
      clearTimeout(t);
    }
  }

  async save(j) {
    await this.rdb.set(key('job', j.id), JSON.stringify(j), 'EX', WEEK);
  }

  async get(id) {
    const b = await this.rdb.get(key('job', id));
    if (b === null) return null;
    return JSON.parse(b);
  }

  // Submit queues a job and returns it with its ticket number.
  async submit(cls, request) {
    if (!Object.hasOwn(CLASSES, cls)) throw new Error('unknown class ' + JSON.stringify(cls));
    const ticket = await this.rdb.incr(key('ticket'));
    const j = { id: newID(), ticket, class: cls, status: 'queued', progress: 'Waiting in the queue', error: null, request, created: now() };
    await this.save(j);
    await this.rdb.zadd(key('queued', cls), ticket, j.id);
    try {
      await this.queues[cls === 'pack' ? PACKS : JOBS].add('ib:job', { id: j.id }, {
        jobId: j.id, priority: CLASSES[cls], attempts: 1,
        removeOnComplete: { age: 24 * 3600 }, removeOnFail: { age: 24 * 3600 },
      });
    } catch (e) {
      await this.rdb.zrem(key('queued', cls), j.id).catch(() => {});
      throw e;
    }
    return j;
  }

  // Position: how many tickets of the same class are queued ahead.
  async position(j) {
    if (j.status !== 'queued') return 0;
    try {
      const r = await this.rdb.zrank(key('queued', j.class), j.id);
      return r === null ? 0 : r;
    } catch (e) {
      return 0;
    }
  }

  // ETA estimates seconds until the job finishes: the jobs ahead plus this
  // one, at the recent average duration, spread over the workers. null when
  // there's no history yet.
  async eta(j, ahead) {
    let durs;
    try { durs = await this.rdb.lrange(key('dur', j.class), 0, 49); } catch (e) { return null; }
    if (!durs.length) return null;
    const avg = durs.reduce((s, d) => s + (parseFloat(d) || 0), 0) / durs.length;
    const w = j.class === 'pack' ? 1 : Math.max(1, this.workers);
    let secs;
    if (j.status === 'queued') {
      const running = await this.rdb.scard(key('running', j.class)).catch(() => 0);
      secs = ((ahead + running) / w + 1) * avg;
    } else if (j.status === 'running') {
      secs = j.started ? avg - (Date.now() - Date.parse(j.started)) / 1000 : 0;
      if (secs < 1) secs = 1;
    } else {
      return null;
    }
    return Math.floor(secs + 0.5);
  }

  // Depths: queued jobs per class.
  async depths() {
    const out = {};
    for (const c of Object.keys(CLASSES).sort()) out[c] = await this.rdb.zcard(key('queued', c)).catch(() => 0);
    return out;
  }

  // serve runs the workers: handler(job, progress) -> result object.
  serve(handler, log = console.log) {
    const work = async (bj) => {
      const j = await this.get(bj.data.id);
      if (!j) return;
      j.status = 'running';
      j.started = now();
      j.progress = 'Starting';
      await this.rdb.zrem(key('queued', j.class), j.id);
      await this.rdb.sadd(key('running', j.class), j.id);
      // Progress is saved in order, and before the final state.
      let saving = this.save(j);
      const progress = (s) => {
        j.progress = s;
        saving = saving.then(() => this.save(j)).catch(() => {});
      };
      let res, err = null, timer;
      try {
        res = await Promise.race([
          handler(j, progress),
          new Promise((_, rej) => { timer = setTimeout(() => rej(new Error('the job took over 2 hours')), TIMEOUT); }),
        ]);
      } catch (e) {
        err = e;
      } finally {
        clearTimeout(timer);
        await this.rdb.srem(key('running', j.class), j.id).catch(() => {});
      }
      await saving;
      j.finished = now();
      if (err) {
        j.status = 'failed';
        j.error = err && err.message ? err.message : String(err);
        j.progress = 'Failed';
      } else {
        j.status = 'done';
        j.result = res;
        j.progress = 'Done';
        const d = (Date.parse(j.finished) - Date.parse(j.started)) / 1000;
        await this.rdb.lpush(key('dur', j.class), d.toFixed(2));
        await this.rdb.ltrim(key('dur', j.class), 0, 49);
      }
      await this.save(j);
      // Failures are reported on the job, not retried.
    };
    const connection = { ...this.conn, maxRetriesPerRequest: null };
    const opts = { connection, prefix: PREFIX, lockDuration: 120000, maxStalledCount: 1 };
    this.running = [
      new Worker(JOBS, work, { ...opts, concurrency: Math.max(1, this.workers) }),
      new Worker(PACKS, work, { ...opts, concurrency: 1 }),
    ];
    for (const w of this.running) w.on('error', (e) => log('workers: ' + e.message));
  }

  async close() {
    await Promise.allSettled(this.running.map((w) => w.close()));
    await Promise.allSettled(Object.values(this.queues).map((q) => q.close()));
    this.rdb.disconnect();
  }
}
